### PR TITLE
CDAP-7676 Custom log processor support in log saver 

### PR DESCRIPTION
**Pull request by gsps1**
Code Changes 
1) Added LogProcessor Interface in cdap-watch-dog API, that external applications can implement for processing logs from log.saver
2) The jar's for log.processor extensions and their properties would be specified through cdap-site.xml similar to how its performed in Authorizer.
3) Refactored AuthorizerInstantiator and AuthorizerClassLoader into ExtensionInstantiatorHelper and ExtensionClassLoader classes in cdap-common for reusability.
4) Test cases to check custom log.processor logic in LogSaverPlugin Test.

Pending
Testing on distributed.

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/lang/ExtensionClassHelper.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+import java.util.Map;
+import java.util.Properties;
+import java.util.jar.Attributes;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+
+/**
+ * Helper class to load and instantiate class from external jar and get extension properties.
+ */
+public final class ExtensionClassHelper {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/ExtensionClassHelper.java:34]: Yet another extension loading mechanism? Can you use the `AbstractExtensionLoader` instead?

Comment (chtyim) [reply cdap-common/src/main/java/co/cask/cdap/common/lang/ExtensionClassHelper.java:34]: Also, please try to minimize creation of util class.

File: cdap-common/src/main/java/co/cask/cdap/common/lang/ExtensionClassLoader.java [hunk 2]
```diff
@@ -30,17 +26,16 @@
 import java.util.Set;
 
 /**
- * {@link DirectoryClassLoader} for {@link Authorizer} extensions.
+ * {@link DirectoryClassLoader} for extensions.
  */
-public class AuthorizerClassLoader extends DirectoryClassLoader {
+public class ExtensionClassLoader extends DirectoryClassLoader {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/ExtensionClassLoader.java:31]: This change is a bit scary. Can you explain the classloading model for the log saver plugin? Do we have a design page?

File: cdap-watchdog-api/src/main/java/co/cask/cdap/api/log/LogProcessor.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+   * Will start receiving messages on log.saver startup
+   *
+   * @param events list of {@link ILoggingEvent}
+   */
+  void process(Iterator<ILoggingEvent> events);
+
+  /**
+   * stop logprocessor
+   */
+  void stop();
```
Comment (chtyim) [inline cdap-watchdog-api/src/main/java/co/cask/cdap/api/log/LogProcessor.java:47]: We better standardize on the naming. We have `initialize()` and `destroy()` for programs.

File: cdap-watchdog-api/src/main/java/co/cask/cdap/api/log/LogProcessor.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * Called during initialize, passed properties for log processor
+   * @param properties
+   */
+  void initialize(Properties properties);
+
+  /**
+   * Process method will be called with iterator of log messages, log messages received will be in sorted order,
+   * sorted by timestamp. This method should not throw any exception, if any unchecked exceptions are thrown,
+   * log.saver will log an error and the processor will not receive messages.
```
Comment (chtyim) [inline cdap-watchdog-api/src/main/java/co/cask/cdap/api/log/LogProcessor.java:37]: Why? Shouldn't the log saver process log the error and proceed?

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/processor/CountingLogProcessor.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      counter++;
+      events.next();
+    }
+  }
+
+  @Override
+  public void stop() {
+    try (FileOutputStream fos = new FileOutputStream(outputFile)) {
+      fos.write(counter);
+      fos.flush();
```
Comment (chtyim) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/processor/CountingLogProcessor.java:54]: close will flush, no need to do it manually.

**Comment by gsps1** [comment]
Closing this as we want to design and decide how to setup the class loader isolation and agree upon API design.

**Pull request closed without merging**